docs(adk): create State and Configuration concept pages (#449)
botpress/docs: adk/concepts/state.mdx

---
title: State
---

> State lets your agent persist data across conversations, users, and the entire bot. The ADK provides three state scopes, each with different lifetimes and access patterns.

Reviewer: ```suggestion
State lets your agent store and reuse data across conversations, users, and the entire bot. The ADK provides three types of states, depending on how long you need the data and who should have access to it.
```

Suggesting a less technical heavy sentence but can reword to fit what you mean

Author: Yes just reworded!

## State scopes

| Scope | Lifetime | Access | Defined in |
|-------|----------|--------|------------|
| Bot | Global, shared across all users and conversations | `bot.state` | `agent.config.ts` |

> | User | Per-user, persists across all of a user's conversations | `user.state` | `agent.config.ts` |

Reviewer: ```suggestion
| User | Per-user, available across all of a user's conversations | `user.state` | `agent.config.ts` |
```

| Conversation | Per-conversation, scoped to a single conversation | Handler's `state` parameter | Conversation definition |

## Defining state schemas

### Bot and user state

Define bot and user state schemas in `agent.config.ts`:

> ```typescript highlight={8-11, 14-18}

Reviewer: Not completely sure what you're highlighting, I think this is off

Author: removed!

  },

  user: {
    state: z.object({
      name: z.string().optional(),
      department: z.string().optional(),
      visitCount: z.number().default(0),
    }),
  },
});
```

<Tip>
  Use `.default()` on your state schema to set initial values. Without defaults, state fields start as `undefined`.
</Tip>

### Conversation state

Conversation state is defined on each Conversation definition using the `state` prop:

```typescript highlight={5-8}
import { Conversation, z } from "@botpress/runtime";

export default new Conversation({
  channel: "*",
  state: z.object({
    topic: z.string().optional(),
    messageCount: z.number().default(0),
  }),
  handler: async ({ execute, state }) => {
    state.messageCount += 1;

    await execute({
      instructions: `You are a helpful assistant. Messages so far: ${state.messageCount}`,
    });
  },
});
```

## Reading and writing state

### Bot state

Import `bot` from `@botpress/runtime` to access bot-level state inside any handler, action, tool, or workflow:

```typescript
import { bot } from "@botpress/runtime";

// Read
const counter = bot.state.ticketCounter;

// Write
bot.state.ticketCounter = counter + 1;
```

### User state

> Import `user` from `@botpress/runtime` to access per-user state inside any handler, action, tool, or workflow:

Reviewer 2: ```suggestion
Import `user` from `@botpress/runtime` to access current-user state inside any handler, action, tool, or workflow:
```

```typescript
import { user } from "@botpress/runtime";

// Read
const name = user.state.name;

// Write
user.state.visitCount = (user.state.visitCount || 0) + 1;
```

### Conversation state

> Conversation state is accessed via the `state` parameter in the handler:

Reviewer 2: ```suggestion
Conversation state is accessed via the `state` parameter in the handler or on a Conversation instance:
```

> ```typescript

Reviewer: can use a highlight here to show what you mean

export default new Conversation({
  channel: "*",
  state: z.object({
    phase: z.string().default("greeting"),
  }),
  handler: async ({ state, execute }) => {
    if (state.phase === "greeting") {
      state.phase = "main";
    }

    await execute({
      instructions: `Current phase: ${state.phase}`,
    });
  },
});
```

<Note>

>   State changes are automatically tracked and persisted. You don't need to call a save method: just mutate the state object directly.

Reviewer: ```suggestion
  State changes are handled automatically. You don't need to call a save method, just update the state object directly.
```

</Note>

## Tags

Tags are string key-value pairs you can attach to bots, users, conversations, messages, and workflows. They're useful for categorization, filtering, and querying.

### Defining tags

Define tag schemas in `agent.config.ts`:

> ```typescript highlight={5-11}

Reviewer: highlight looks off

Author: fixed

export default defineConfig({
  name: "my-agent",

  bot: {
    tags: {
      environment: { title: "Environment", description: "Current deployment environment" },
    },
  },

  user: {
    tags: {
      role: { title: "Role", description: "User's role in the organization" },
    },
  },

  conversation: {
    tags: {
      priority: { title: "Priority", description: "Conversation priority level" },
    },
  },
});
```

### Reading and writing tags

> 

Reviewer 3: try to always include some kind of lead-in or explanation before presenting an example or code snippet, no matter how self-explanatory the example is.

this helps readers quickly scan for the information they need on a page with a dozen or more code snippets, especially in this case where the snippet contains example variable tag names that are less likely to draw the reader's eye.

Author: Good point! I added a lead-in sentence before the tags example

```typescript
import { bot, user } from "@botpress/runtime";

// Bot tags
bot.tags.environment = "production";
const env = bot.tags.environment;

// User tags
user.tags.role = "admin";
```

> Conversation tags are accessed via the `conversation` instance in handlers:

Reviewer 2: ```suggestion
Conversation tags are accessed via a `conversation` instance:
```

since you can also do `Conversation.get(<id>)`

> Conversation tags are accessed via the `conversation` instance in handlers:

Reviewer 3: **nit:** try to use [active voice](https://developers.google.com/tech-writing/one/active-voice) when possible. who accesses the tags?
```suggestion
You can access conversation tags via the `conversation` instance.
```

```typescript
export default new Conversation({
  channel: "*",
  handler: async ({ conversation, execute }) => {
    // Read
    const priority = conversation.tags.priority;

    // Write
    conversation.tags.priority = "high";
  },
});
```

<Note>

>   System tags (set by integrations, containing `:` in the key) are read-only. You can read them but any writes are silently ignored.

Reviewer: I think an example code would help here instead of using a note

Author: added

</Note>

## State references

> You can store references to workflow instances in state. They are automatically serialized when saved and loaded back as full instances when read:

Reviewer: ```suggestion
You can store workflow instances in state. They are saved automatically and loaded back as full instances when read:
```
Unless serialized is an important term, this just needs to be reworded

Author: yes

> ```typescript

Reviewer: instead of commenting in the code, can highlight whats relevant

Author: added!
